### PR TITLE
SEP-10: Clarifying Participants

### PR DESCRIPTION
**Pull request by JakeUrban**
This change does not change any functionality related to SEP-10. Rather, it better defines the participants involved in the process and updating the document to always reference those participant definitions. This will help avoid confusion when referencing the participants, instead of using general terms like `client` and `server`.

If merged, I'll update #760 to use the term **Client Server**

### CODE REVIEW

File: ecosystem/sep-0010.md
Comment (leighmcculloch) [inline ecosystem/sep-0010.md:25]: 💡 I think `owner` is okay, but `account holder` is more specific and descriptive, and also a common phrase used in financial products which helps provide more depth here.
```suggestion
- The **Client Application**: the application used by the account holder of the **Authenticating Account**.
```

File: ecosystem/sep-0010.md
Comment (leighmcculloch) [inline ecosystem/sep-0010.md:22]: 🔠 I think we probably want to keep `The` => `A` for all of these, because we're not referring to a definite article, but an indefinite article. We're referring to any toml server participating in SEP-10.

File: ecosystem/sep-0010.md
Comment (leighmcculloch) [inline ecosystem/sep-0010.md:23]: 💡  I think TOML Server and Web Auth Server are being a little too prescriptive and distancing the definition of the participants from what matters for SEP-10. SEP-10 doesn't need you to have these servers in the modern sense of the word, but needs you to provides:
- A home domain that has a toml file at a URL at that domain.
- An web auth endpoint.

Whatever approach an implementer takes to providing those is out-of-scope, and making the home domain a participant is critical to the rest of the specification where the home domain is a thing.

```suggestion
- A **Home Domain**: a domain hosting a [SEP-1 stellar.toml](sep-0001.md) containing a `WEB_AUTH_ENDPOINT` (URL) and `SIGNING_KEY` (`G...`).
- A **Web Auth Endpoint**: an endpoint providing the GET and POST operations discussed in this document. The endpoint may be hosted on the **Home Domain**, a sub-domain of the **Home Domain**, or any other domain. 
- A **Web Auth Domain**: the domain hosting the **Web Auth Endpoint**.
```

Thoughts?

**Review by leighmcculloch — COMMENTED** [review]
Some thoughts below just on the definitions, I didn't read the full PR yet, but can do so when we agree on if we prefer TOML Server / Web Auth Server, or Home Domain and Web Auth Endpoint,

File: ecosystem/sep-0010.md
Comment (JakeUrban) [reply ecosystem/sep-0010.md:23]: I like the change from `TOML Server` to `Home Domain` because you're right, we use the word 'home domain' throughout the doc and in request parameters. 

However I think the change from `Web Auth Server` to `Web Auth Domain` & `Web Auth Endpoint` is a bit unnecessary. Unlike the home domain hosting a static file at a path, an actual API server is required to process authentication requests. I don't see the value of trying to de-couple the endpoint and domain of the web auth server.

Comment (leighmcculloch) [reply ecosystem/sep-0010.md:23]: I see your point. How about we replace Web Auth Endpoint with Web Auth Server, but keep Web Auth Domain in the list because it is an important attribute elsewhere in the SEP?

Comment (JakeUrban) [reply ecosystem/sep-0010.md]: Hmm if the goal is to define a component for the `web_auth_domain` key mentioned in the SEP, then I think we should go with what you had: `Web Auth Endpoint` & `Web Auth Domain`, since the TOML uses `WEB_AUTH_ENDPOINT`.

However, I'm not sure it worth it to explicitly define a `Web Auth Domain`, since there are only two places where it would be used. Do you think the current phrasing is confusing?

(in the abstract)
> 1. The **Client Application** verifies that if the transaction has a Manage Data operation with key `web_auth_domain` that it has:
>       1. Source account set to the ``SIGNING_KEY`` from the **Home Domain**'s [SEP-1 stellar.toml](sep-0001.md).
>       1. Value set to the **Web Auth Server**'s domain.

(in the response)
> * `manage_data(source: server_account, key: 'web_auth_domain', value: web_auth_domain)`
>        * The source account is the **Home Domain**'s ``SIGNING_KEY``, the value is the domain of the **Web Auth Server**. It can be at most 64 characters.

Comment (leighmcculloch) [reply ecosystem/sep-0010.md]: I see your point. It's not anymore ambiguous to refer to the domain of the Web Auth Server.

I also agree for Web Auth Endpoint so as to remain consistent with the stellar.toml, so there is one less new concept and we continue to reuse the existing concept.

Comment (JakeUrban) [reply ecosystem/sep-0010.md]: Ok, so the take-away is to change `Web Auth Server` to `Web Auth Endpoint`, but not use the additional `Web Auth Domain` term, since we can easily say `the domain of the **Web Auth Endpoint**`?

Comment (leighmcculloch) [reply ecosystem/sep-0010.md]: Yup, that's how I interpreted what you were saying and I think that strikes the balance.

File: ecosystem/sep-0010.md
Comment (leighmcculloch) [inline ecosystem/sep-0010.md:22]: 💡 I don't think you need this, it is rather self explanatory especially given the proximity to the next line.
```suggestion
- A **Home Domain**: a domain hosting a [SEP-1 stellar.toml](sep-0001.md) containing a `WEB_AUTH_ENDPOINT` (URL) and `SIGNING_KEY` (`G...`).
```

File: ecosystem/sep-0010.md
Comment (leighmcculloch) [inline ecosystem/sep-0010.md:23]: ❗ The endpoint isn't a domain, but it leads by saying that it is a domain.

```suggestion
- A **Web Auth Endpoint**: an endpoint providing the GET and POST operations discussed in this document. The endpoint may be hosted on the **Home Domain**, a sub-domain of the **Home Domain**, or any other domain.
```

❗ After re-reading this I think it is still necessary to either keep the sentence `The domain where the endpoint is hosted is referred to as the web auth domain.` which is in the original, or to add that statement to a new list item. A list item would be more explicit:

```suggestion
- A **Web Auth Endpoint**: an endpoint providing the GET and POST operations discussed in this document. The endpoint may be hosted on a **Home Domain**, a sub-domain of a **Home Domain**, or any other domain.
- A **Web Auth Domain**: the domain where the **Web Auth Endpoint** is hosted. 
```

File: ecosystem/sep-0010.md
Comment (leighmcculloch) [inline ecosystem/sep-0010.md:25]: ❗ We should be consistent in this list with the A vs The leads:
```suggestion
- A **Client Application**: the software used by the account holder of the **Authenticating Account**.
```

File: ecosystem/sep-0010.md
Comment (leighmcculloch) [inline ecosystem/sep-0010.md:24]: ❗ The component is the account not the public key. In the API we might reference the public key but the component in play is the account itself.
```suggestion
- An **Authenticating Account**: a Stellar account being authenticated.
```

File: ecosystem/sep-0010.md
Comment (JakeUrban) [reply ecosystem/sep-0010.md]: I agree, its ultimately better to be explicit even if `Web Auth Domain` is seldom used. Editing your 2nd suggestion so that it doesn't conflict with the first -- I'm committing both.

File: ecosystem/sep-0010.md
Comment (JakeUrban) [reply ecosystem/sep-0010.md:24]: I don't feel strongly either way, but the account may not actually exist, which is why I chose to mention the public key specifically.

File: ecosystem/sep-0010.md
Comment (leighmcculloch) [inline ecosystem/sep-0010.md:89]: ❗️In the last sentence of this section it called out that the client must verify the servers signature. That's been removed, but it seems unnecessary to me to remove it if this change is scoped to clarifying the participants. We kept the statement to verify the home domain, which seems like we're playing more importance on that when everything that needs verifying is important.

File: ecosystem/sep-0010.md
Comment (leighmcculloch) [inline ecosystem/sep-0010.md:120]: 💡 I think it is worth adding a dot point here above the key dot point, that states that the source account is the `**Authenticating Account**` so you make use of the participant bold text here.

File: ecosystem/sep-0010.md
Comment (leighmcculloch) [inline ecosystem/sep-0010.md:124]: 💡 I think it is worth splitting the source account, and value into separate dot points.

File: ecosystem/sep-0010.md
Comment (leighmcculloch) [inline ecosystem/sep-0010.md:163]: 💡 It's unnecessary for the server to verify this because they signed the transaction. Our SDKs might do it, but there's little need in the specification to specify it. It doesn't hurt to specify it, but this is why we don't specify verifying the web auth domain here as well.

**Review by leighmcculloch — COMMENTED** [review]
I've got some more changes I'd like to make. Sorry to bouncing on some of these things, but there's a lot of content, subtlety and words and semantics here. I've outlined some requested suggestions below. I've also opened a PR making the suggestions: https://github.com/JakeUrban/stellar-protocol/pull/1. I'm happy to talk synchronously through these changes if that would be helpful too.

---

👏🏻 I'm not sure I said this yet, but this PR goes a long way to clearing up ambiguities and making the document more mature. Thanks for identifying that this was needed and taking charge of it.

File: ecosystem/sep-0010.md
Comment (JakeUrban) [reply ecosystem/sep-0010.md:48]: I think having a `Web Auth Endpoint`, `Web Auth Server`, and `Web Auth Domain` adds more potential for confusion than having a single `Web Auth` participant and referencing related properties of that participant. Listing each as participants makes it seem like they are independent of each other, even if the definitions at the top of the document clarify that they're all related.

My preference would be to only define `Web Auth Server` in the participants list, but as you suggest we can make sure we're referring to the `Web Auth Server`'s endpoint when appropriate, and same for it's domain.

Comment (leighmcculloch) [reply ecosystem/sep-0010.md:48]: I see your point. 👍

I think this also raises the question if the authenticating account is necessary, given that it is just a component of the client and it might be better to return to Client and Server, but do as you've done here and make them explicit identifiers by holding them and using them intentionally. Then attribution introduces only one new concept, the Client Server.

Thoughts?

Comment (JakeUrban) [reply ecosystem/sep-0010.md:48]: So you're suggesting to have the following components defined in the abstract:
- Home Domain
- Client Application
- Web Auth Server

But still use the bolded style when referring to properties related to the components defined in the abstract? I think we should reserve the bolded style to explicitly defined components, because bolding effectively implies that they are defined.

Do you think any of the following phrases are ambiguous?

> Source account set to the account the **Client Application** is authenticating

> The **Client Application** obtains a unique [`challenge`](#challenge) from the endpoint implemented by the **Web Auth Server**

> The source account is the **Home Domain**'s ``SIGNING_KEY``, the value is the domain on which the **Web Auth Server** is hosted.

The above phrases replaces usage of **Authenticating Account**, **Web Auth Endpoint**, and **Web Auth Domain** terms. It may be more verbose but I don't think there is potential for misunderstanding.

Client Attribution would then introduce the **Client Server**, along with language like the following:

> The **Client Server** should implement an endpoint that accepts a challenge transaction from the **Client Application** and signs it using the `SIGNING_KEY` listed on the [SEP-1 stellar.toml](...) file hosted on it's domain

Comment (leighmcculloch) [reply ecosystem/sep-0010.md:48]: I was meaning what you outlined, only three bolded components and rely on as brief identification for others, then the client server becomes a new component.

I'd like to find a more succinct form for the server account/signing key though. 

I think if we rename Web Auth Server to simply Server, then to say Client account and Server account is sufficient, then Client Server is still distinct?

Comment (JakeUrban) [reply ecosystem/sep-0010.md:48]: Sounds good to me!

File: ecosystem/sep-0010.md
Comment (JakeUrban) [reply ecosystem/sep-0010.md:165]: Aren't all verifications above this comment unnecessary though? The only thing the server needs to do is verify that it's account signed the transaction and then verify the other signatures depending on whether the client account exists. Should I remove all verification points above this comment?

Comment (leighmcculloch) [reply ecosystem/sep-0010.md:165]: You're right, but I don't think there is significant value in removing existing statements made here, I just wouldn't go out of your way to add new ones.

File: ecosystem/sep-0010.md
Comment (leighmcculloch) [inline ecosystem/sep-0010.md:26]: 🔠 
```suggestion
    - The account being authenticated is the **Client Account**.
```

File: ecosystem/sep-0010.md
Comment (leighmcculloch) [inline ecosystem/sep-0010.md:24]: 🔠 
```suggestion
    - The `SIGNING_KEY` from the **Home Domain** is the **Server Account**.
```

File: ecosystem/sep-0010.md
Comment (leighmcculloch) [inline ecosystem/sep-0010.md:23]: 💡 This is sort of out-of-scope of your PR but I think this tiny tweak is worthwhile. The 'any other domain' doesn't quite communicate the same intent as saying it can be a different domain, although they're similar.
```suggestion
- A **Server**: a server providing the `WEB_AUTH_ENDPOINT` that implements the GET and POST operations discussed in this document. The server's domain may be the **Home Domain**, a sub-domain of the **Home Domain**, or a different domain.
```

File: ecosystem/sep-0010.md
Comment (leighmcculloch) [inline ecosystem/sep-0010.md:31]: 🎉 This referencing here of both the signing key and the server account at the first mention is helpful, great idea.

File: ecosystem/sep-0010.md
Comment (leighmcculloch) [inline ecosystem/sep-0010.md:44]: ❗ I think we need to keep the text that was removed. It is important that the client verify it matches the domain they requested from. I don't want that distinction to be lost.
```suggestion
   1. Value set to the **Server**'s domain that the client requested the challenge from.
```

File: ecosystem/sep-0010.md
Comment (leighmcculloch) [inline ecosystem/sep-0010.md:129]: ❗ I think this still needs to say that the intention is to check what the server is using, not the home domain. If the client wanted to check what the home domain used, it'd look at the toml.
```suggestion
* `network_passphrase`: (optional but recommended) Stellar network passphrase used by the **Server**. This allows a **Client** to verify that it's using the correct passphrase when signing and is useful for identifying when a **Client** or **Server** have been configured incorrectly.
```

**Review by leighmcculloch — COMMENTED** [review]
This is awesome and so much easier to read with these changes. I've got a two small things that needs tweaking (❗) before merging, and three small suggestions for wording (💡🔠) that I'll defer to you but I think read clearer and more direct.

Mention me when these are addressed and I'll approve.

**Comment by JakeUrban** [comment]
@leighmcculloch no issues with your suggestions, I committed them all 👍

**Review by leighmcculloch — APPROVED** [review]
Looks great!

Before merging, can you update the meta data? Version 3.1.2.

**Pull request merged**
